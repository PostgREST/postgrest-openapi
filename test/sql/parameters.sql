-- shows common query parameters
select *
from jsonb_each(postgrest_openapi_spec('{test}')->'components'->'parameters')
where key in ('select', 'order', 'limit', 'offset', 'on_conflict', 'columns', 'or', 'and', 'not.or', 'not.and');

-- shows common headers
select key, jsonb_pretty(value)
from jsonb_each(postgrest_openapi_spec('{test}')->'components'->'parameters')
where key in ('preferGet', 'preferPost', 'preferPut', 'preferPatch', 'preferDelete', 'preferPostRpc', 'range');

-- shows table columns as parameters
select key, jsonb_pretty(value)
from jsonb_each(postgrest_openapi_spec('{test}')->'components'->'parameters')
where key like 'rowFilter.products.%';

-- shows view columns as parameters
select key, jsonb_pretty(value)
from jsonb_each(postgrest_openapi_spec('{test}')->'components'->'parameters')
where key like 'rowFilter.big_products.%';
