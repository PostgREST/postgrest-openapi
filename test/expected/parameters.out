-- shows common query parameters
select *
from jsonb_each(get_postgrest_openapi_spec('{test}')->'components'->'parameters')
where key in ('select', 'order', 'limit', 'offset', 'on_conflict', 'columns', 'or', 'and', 'not.or', 'not.and');
     key     |                                                                                     value                                                                                     
-------------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 or          | {"in": "query", "name": "or", "schema": {"type": "string"}, "explode": false, "description": "Logical operator to combine filters using OR"}
 and         | {"in": "query", "name": "and", "schema": {"type": "string"}, "explode": false, "description": "Logical operator to combine filters using AND (the default for query params)"}
 limit       | {"in": "query", "name": "limit", "schema": {"type": "integer"}, "explode": false, "description": "Limit the number of rows returned"}
 order       | {"in": "query", "name": "order", "schema": {"type": "array", "items": {"type": "string"}}, "explode": false, "description": "Ordering by column"}
 not.or      | {"in": "query", "name": "not.or", "schema": {"type": "string"}, "explode": false, "description": "Negate the logical operator to combine filters using OR"}
 offset      | {"in": "query", "name": "offset", "schema": {"type": "integer"}, "explode": false, "description": "Skip a certain number of rows"}
 select      | {"in": "query", "name": "select", "schema": {"type": "array", "items": {"type": "string"}}, "explode": false, "description": "Vertical filtering of columns"}
 columns     | {"in": "query", "name": "columns", "schema": {"type": "string"}, "explode": false, "description": "Specify which keys from the payload will be inserted"}
 not.and     | {"in": "query", "name": "not.and", "schema": {"type": "string"}, "explode": false, "description": "Negate the logical operator to combine filters using AND"}
 on_conflict | {"in": "query", "name": "on_conflict", "schema": {"type": "string"}, "explode": false, "description": "Columns that resolve the upsert conflict"}
(10 rows)

-- shows common headers
select key, jsonb_pretty(value)
from jsonb_each(get_postgrest_openapi_spec('{test}')->'components'->'parameters')
where key in ('preferGet', 'preferPost', 'preferPut', 'preferPatch', 'preferDelete', 'preferPostRpc', 'range');
      key      |                                 jsonb_pretty                                 
---------------+------------------------------------------------------------------------------
 range         | {                                                                           +
               |     "in": "header",                                                         +
               |     "name": "Range",                                                        +
               |     "schema": {                                                             +
               |         "type": "string"                                                    +
               |     },                                                                      +
               |     "example": "0-4",                                                       +
               |     "description": "For limits and pagination"                              +
               | }
 preferGet     | {                                                                           +
               |     "in": "header",                                                         +
               |     "name": "Prefer",                                                       +
               |     "schema": {                                                             +
               |         "allOf": [                                                          +
               |             {                                                               +
               |                 "$ref": "#/components/schemas/header.preferHandling"        +
               |             },                                                              +
               |             {                                                               +
               |                 "$ref": "#/components/schemas/header.preferTimezone"        +
               |             },                                                              +
               |             {                                                               +
               |                 "$ref": "#/components/schemas/header.preferCount"           +
               |             }                                                               +
               |         ]                                                                   +
               |     },                                                                      +
               |     "explode": true,                                                        +
               |     "examples": {                                                           +
               |         "all": {                                                            +
               |             "value": {                                                      +
               |                 "count": "",                                                +
               |                 "handling": "lenient",                                      +
               |                 "timezone": ""                                              +
               |             },                                                              +
               |             "summary": "All default preferences"                            +
               |         },                                                                  +
               |         "nothing": {                                                        +
               |             "summary": "No preferences"                                     +
               |         }                                                                   +
               |     },                                                                      +
               |     "description": "Specify a required or optional behavior for the request"+
               | }
 preferPut     | {                                                                           +
               |     "in": "header",                                                         +
               |     "name": "Prefer",                                                       +
               |     "schema": {                                                             +
               |         "allOf": [                                                          +
               |             {                                                               +
               |                 "$ref": "#/components/schemas/header.preferHandling"        +
               |             },                                                              +
               |             {                                                               +
               |                 "$ref": "#/components/schemas/header.preferTimezone"        +
               |             },                                                              +
               |             {                                                               +
               |                 "$ref": "#/components/schemas/header.preferReturn"          +
               |             },                                                              +
               |             {                                                               +
               |                 "$ref": "#/components/schemas/header.preferCount"           +
               |             },                                                              +
               |             {                                                               +
               |                 "$ref": "#/components/schemas/header.preferTx"              +
               |             }                                                               +
               |         ]                                                                   +
               |     },                                                                      +
               |     "explode": true,                                                        +
               |     "examples": {                                                           +
               |         "all": {                                                            +
               |             "value": {                                                      +
               |                 "tx": "commit",                                             +
               |                 "count": "",                                                +
               |                 "return": "minimal",                                        +
               |                 "handling": "lenient",                                      +
               |                 "timezone": ""                                              +
               |             },                                                              +
               |             "summary": "All default preferences"                            +
               |         },                                                                  +
               |         "nothing": {                                                        +
               |             "summary": "No preferences"                                     +
               |         }                                                                   +
               |     },                                                                      +
               |     "description": "Specify a required or optional behavior for the request"+
               | }
 preferPost    | {                                                                           +
               |     "in": "header",                                                         +
               |     "name": "Prefer",                                                       +
               |     "schema": {                                                             +
               |         "allOf": [                                                          +
               |             {                                                               +
               |                 "$ref": "#/components/schemas/header.preferHandling"        +
               |             },                                                              +
               |             {                                                               +
               |                 "$ref": "#/components/schemas/header.preferTimezone"        +
               |             },                                                              +
               |             {                                                               +
               |                 "$ref": "#/components/schemas/header.preferReturn"          +
               |             },                                                              +
               |             {                                                               +
               |                 "$ref": "#/components/schemas/header.preferCount"           +
               |             },                                                              +
               |             {                                                               +
               |                 "$ref": "#/components/schemas/header.preferResolution"      +
               |             },                                                              +
               |             {                                                               +
               |                 "$ref": "#/components/schemas/header.preferMissing"         +
               |             },                                                              +
               |             {                                                               +
               |                 "$ref": "#/components/schemas/header.preferTx"              +
               |             }                                                               +
               |         ]                                                                   +
               |     },                                                                      +
               |     "explode": true,                                                        +
               |     "examples": {                                                           +
               |         "all": {                                                            +
               |             "value": {                                                      +
               |                 "tx": "commit",                                             +
               |                 "count": "",                                                +
               |                 "return": "minimal",                                        +
               |                 "missing": "null",                                          +
               |                 "handling": "lenient",                                      +
               |                 "timezone": "",                                             +
               |                 "resolution": ""                                            +
               |             },                                                              +
               |             "summary": "All default preferences"                            +
               |         },                                                                  +
               |         "nothing": {                                                        +
               |             "summary": "No preferences"                                     +
               |         }                                                                   +
               |     },                                                                      +
               |     "description": "Specify a required or optional behavior for the request"+
               | }
 preferPatch   | {                                                                           +
               |     "in": "header",                                                         +
               |     "name": "Prefer",                                                       +
               |     "schema": {                                                             +
               |         "allOf": [                                                          +
               |             {                                                               +
               |                 "$ref": "#/components/schemas/header.preferHandling"        +
               |             },                                                              +
               |             {                                                               +
               |                 "$ref": "#/components/schemas/header.preferTimezone"        +
               |             },                                                              +
               |             {                                                               +
               |                 "$ref": "#/components/schemas/header.preferReturn"          +
               |             },                                                              +
               |             {                                                               +
               |                 "$ref": "#/components/schemas/header.preferCount"           +
               |             },                                                              +
               |             {                                                               +
               |                 "$ref": "#/components/schemas/header.preferTx"              +
               |             },                                                              +
               |             {                                                               +
               |                 "$ref": "#/components/schemas/header.preferMaxAffected"     +
               |             }                                                               +
               |         ]                                                                   +
               |     },                                                                      +
               |     "explode": true,                                                        +
               |     "examples": {                                                           +
               |         "all": {                                                            +
               |             "value": {                                                      +
               |                 "tx": "commit",                                             +
               |                 "count": "",                                                +
               |                 "return": "minimal",                                        +
               |                 "handling": "lenient",                                      +
               |                 "timezone": "",                                             +
               |                 "max-affected": ""                                          +
               |             },                                                              +
               |             "summary": "All default preferences"                            +
               |         },                                                                  +
               |         "nothing": {                                                        +
               |             "summary": "No preferences"                                     +
               |         }                                                                   +
               |     },                                                                      +
               |     "description": "Specify a required or optional behavior for the request"+
               | }
 preferDelete  | {                                                                           +
               |     "in": "header",                                                         +
               |     "name": "Prefer",                                                       +
               |     "schema": {                                                             +
               |         "allOf": [                                                          +
               |             {                                                               +
               |                 "$ref": "#/components/schemas/header.preferHandling"        +
               |             },                                                              +
               |             {                                                               +
               |                 "$ref": "#/components/schemas/header.preferTimezone"        +
               |             },                                                              +
               |             {                                                               +
               |                 "$ref": "#/components/schemas/header.preferReturn"          +
               |             },                                                              +
               |             {                                                               +
               |                 "$ref": "#/components/schemas/header.preferCount"           +
               |             },                                                              +
               |             {                                                               +
               |                 "$ref": "#/components/schemas/header.preferTx"              +
               |             },                                                              +
               |             {                                                               +
               |                 "$ref": "#/components/schemas/header.preferMaxAffected"     +
               |             }                                                               +
               |         ]                                                                   +
               |     },                                                                      +
               |     "explode": true,                                                        +
               |     "examples": {                                                           +
               |         "all": {                                                            +
               |             "value": {                                                      +
               |                 "tx": "commit",                                             +
               |                 "count": "",                                                +
               |                 "return": "minimal",                                        +
               |                 "handling": "lenient",                                      +
               |                 "timezone": "",                                             +
               |                 "max-affected": ""                                          +
               |             },                                                              +
               |             "summary": "All default preferences"                            +
               |         },                                                                  +
               |         "nothing": {                                                        +
               |             "summary": "No preferences"                                     +
               |         }                                                                   +
               |     },                                                                      +
               |     "description": "Specify a required or optional behavior for the request"+
               | }
 preferPostRpc | {                                                                           +
               |     "in": "header",                                                         +
               |     "name": "Prefer",                                                       +
               |     "schema": {                                                             +
               |         "allOf": [                                                          +
               |             {                                                               +
               |                 "$ref": "#/components/schemas/header.preferHandling"        +
               |             },                                                              +
               |             {                                                               +
               |                 "$ref": "#/components/schemas/header.preferTimezone"        +
               |             },                                                              +
               |             {                                                               +
               |                 "$ref": "#/components/schemas/header.preferCount"           +
               |             },                                                              +
               |             {                                                               +
               |                 "$ref": "#/components/schemas/header.preferTx"              +
               |             },                                                              +
               |             {                                                               +
               |                 "$ref": "#/components/schemas/header.preferParams"          +
               |             }                                                               +
               |         ]                                                                   +
               |     },                                                                      +
               |     "explode": true,                                                        +
               |     "examples": {                                                           +
               |         "all": {                                                            +
               |             "value": {                                                      +
               |                 "tx": "commit",                                             +
               |                 "count": "",                                                +
               |                 "params": "",                                               +
               |                 "handling": "lenient",                                      +
               |                 "timezone": ""                                              +
               |             },                                                              +
               |             "summary": "All default preferences"                            +
               |         },                                                                  +
               |         "nothing": {                                                        +
               |             "summary": "No preferences"                                     +
               |         }                                                                   +
               |     },                                                                      +
               |     "description": "Specify a required or optional behavior for the request"+
               | }
(7 rows)

